### PR TITLE
Added support for hint index for mysql DB driver

### PR DESCRIPTION
**Pull request by fdezmc**
Allowing support for hint index in MySQL with ActiveQuery and ActiveRecord (#10869).

Currently hint index with Expression for ActiveQuery with joins fails: 
(_yiiunit\framework\db\mysql\ActiveQueryTest::testFromIndexHint is incorrect_)
```
$query = (new Query)
    ->from([new Expression('{{user}} {{t}} FORCE INDEX (primary) IGNORE INDEX FOR ORDER BY (i1)')])
    ->leftJoin(['p' => 'profile'], 'user.id = profile.user_id USE INDEX (i2)');

--- Expected
+++ Actual
@@ @@
-'SELECT * FROM {{user}} {{t}} FORCE INDEX (primary) IGNORE INDEX FOR ORDER BY (i1) LEFT JOIN `profile` `p` USE INDEX (i2) ON user.id = profile.user_id'
+'SELECT * FROM {{user}} {{t}} FORCE INDEX (primary) IGNORE INDEX FOR ORDER BY (i1) LEFT JOIN `profile` `p` ON user.id = profile.user_id USE INDEX (i2)'
```

and hint index with ActiveRecord is not currently supported.

***This pull request allow all functionality for hint index on mysql DB driver***

ActiveQuery with no joins:
```
$query = (new Query)
	->from(['o' => order])->hintIndex(['order', 'forceIndex', 'primary']);

+++ Actual
@@ @@
'SELECT `order`.* FROM `order` FORCE INDEX (primary)'
```

ActiveQuery with joins:
```
$query = (new Query)
    ->from('order')->hintIndex([['order', 'forceIndex', 'primary'], ['order', 'ignoreIndex', 'FK_order_customer_id', 'orderBy']])
    ->leftJoin('customer', 'order.customer_id = customer.id')->hintIndex(['customer', 'useIndex', 'primary']);

+++ Actual
@@ @@
'SELECT * FROM `order` FORCE INDEX (`primary`) IGNORE INDEX FOR ORDER BY (`FK_order_customer_id`) LEFT JOIN `customer` USE INDEX (`primary`) ON order.customer_id = customer.id
```

ActiveRecord with no joins:
```
$hintIndex = ['order', 'useIndex', ['primary', 'FK_order_customer_id']];
$query = Order::find()->hintIndex($hintIndex)->where(['id' => 1, 'customer_id' => 1]);
+++ Actual
@@ @@
'SELECT * FROM `order` USE INDEX (`primary`, `FK_order_customer_id`) WHERE (`id`=1) AND (`customer_id`=1)'
```

ActiveRecord with joins:
```
$hintIndex = ['order', 'forceIndex', 'primary'];
$query = Order::find()->hintIndex($hintIndex)
	->leftJoin('customer', 'order.customer_id = customer.id')->hintIndex(['customer', 'forceIndex', 'primary']);
+++ Actual
@@ @@
'SELECT `order`.* FROM `order` FORCE INDEX (`primary`) LEFT JOIN `customer` FORCE INDEX (`primary`) ON order.customer_id = customer.id'
```

Hinted columns as string:
```
$hintIndex = ['order', 'ignoreIndex', 'FK_order_customer_id'];
$query = Order::find()->hintIndex($hintIndex)->where(['id' => 1, 'customer_id' => 1]);
```

Multiple hints at once
```
$hintIndex = [['order', 'forceIndex', ['primary']], ['order', 'ignoreIndex', 'FK_order_customer_id', 'orderBy']];
$query = Order::find()->hintIndex($hintIndex)->orderBy('customer_id');
```

With alias
```
$hintIndex = ['order', 'ignoreIndex', 'customer_id'];
$query = Order::find()->from(['o' => 'order'])->hintIndex($hintIndex)->where(['id' => 1, 'customer_id' => 1]);
```

With joinWith, primary and alias
```
$hintIndex = ['order', 'forceIndex', 'primary'];
$query = Order::find()->hintIndex($hintIndex)->joinWith('items');
```

With relation by puntuation and hint on join table
```
$hintIndex = ['item', 'forceIndex', 'primary', 'orderBy'];
$query = Order::find()->joinWith('orderItems.item')->hintIndex($hintIndex)->orderBy('item.id');
```

All logic is in the specific DB driver class:
- _/yii2/framework/db/mysql/QueryBuilder.php_
- _/yii2/framework/db/sqlite/QueryBuilder.php_ (*)
- _/yii2/framework/db/cubrid/QueryBuilder.php_ (*)
- _/yii2/framework/db/mssql/QueryBuilder.php_ (*)
- _/yii2/framework/db/oci/QueryBuilder.php_ (*)
- _/yii2/framework/db/pgsql/QueryBuilder.php_ (*)

_(*) This DB drivers support for hint index is not implemente in this pull-request yet. Query::hintIndex property will be ignore_

I have no experience enought with other DB drivers than mysql, so just write functionality for mysql.
Support for the left DB drivers can be done following this approach.

### CODE REVIEW

File: framework/db/QueryBuilder.php [hunk 1]
```diff
@@ -775,12 +775,14 @@ public function buildSelect($columns, &$params, $distinct = false, $selectOption
     }
 
     /**
-     * @param array $tables
+     * @param Query $query the [[Query]] object from which the SQL statement will be generated.
      * @param array $params the binding parameters to be populated
      * @return string the FROM clause built from [[Query::$from]].
      */
-    public function buildFrom($tables, &$params)
+    public function buildFrom($query, &$params)
```
Comment (samdark) [inline framework/db/QueryBuilder.php]: Looks like potential backwards compatibility break.

Comment (fdezmc) [inline framework/db/QueryBuilder.php]: Fixed backwards compatibility, but travis-ci through this error:

> The command "tests/data/travis/apc-setup.sh" failed and exited with 1 during .

Has this relation with my commit? All test related with mysql hint index are passed on my local machine.

**Comment by fdezmc** [comment]
Fixed backwards compatibility.
All test related with mysql hint index are passed on my local machine.

travis-ci through this error for php 5.4, 5.5, 5.6 versions:

> No releases available for package "pecl.php.net/apc"
> install failed
> The command "tests/data/travis/apc-setup.sh" failed and exited with 1 during .
> Your build has been stopped.

Is there any problem with travis-ci?

**Comment by SilverFire** [comment]
Hinting for other supported DBMSs look different.

- [Oracle optimizer hints](http://docs.oracle.com/cd/B19306_01/server.102/b14211/hintsref.htm#i17496)
- [PgSQL as a separate extension](https://osdn.net/projects/pghintplan/wiki/FrontPage#h2-.E6.A9.9F.E8.83.BD.E5.88.B6.E9.99.90)
- [Microsoft SQL](https://msdn.microsoft.com/en-us/library/ms181714.aspx#Examples)

The enhancement should be compatible with these DBMS

**Comment by fdezmc** [comment]
Hi @SilverFire, the enhancement and the pull-request are compatible with other DBMS.
Unfortunately only support for MySQL DB driver is implemented in this pull-request. This is because of my inexperience with other DBMS, so drivers different than MySQL will ignore the `Query::hintIndex` property.

Anyway following this approach, anyone with enough knowledge, can pull-request support for pending DBMS by editing specific DB driver QueryBuilder class. 
Basically, `QueryBuilder::buildFrom` and `QueryBuilder::buildJoin` method receive the `Query` object as param which can be used to rewrite the SQL sentence to fit the specific DB driver hint index implementations.
